demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+#ifndef DEMO_HTTP_UPLOAD_DATA_SIZE
+    #define DEMO_HTTP_UPLOAD_DATA_SIZE    ( sizeof( DEMO_HTTP_UPLOAD_DATA ) - 1 )
+#endif
+
+/**
+ * @brief A buffer used in the demo for storing HTTP request headers and
+ * HTTP response headers and body.
+ *
+ * @note This demo shows how the same buffer can be re-used for storing the HTTP
+ * response after the HTTP request is sent out. However, the user can also
+ * decide to use separate buffers for storing the HTTP request and response.
+ */
+static uint8_t userBuffer[ USER_BUFFER_LENGTH ];
+
+/**
+ * @brief Represents header data that will be sent in an HTTP request.
+ */
+static HTTPRequestHeaders_t requestHeaders;
+
+/**
+ * @brief Configurations of the initial request headers that are passed to
+ * #HTTPClient_InitializeRequestHeaders.
+ */
+static HTTPRequestInfo_t requestInfo;
+
+/**
+ * @brief Represents a response returned from an HTTP server.
+ */
+static HTTPResponse_t response;
+
+/* The host address string extracted from the pre-signed URL. */
+static char serverHost;
+
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Retrieve the size of the S3 object that is specified in pPath.
+ *
+ * @param[out] pFileSize - The size of the S3 object.
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ */
+static HTTPStatus_t getS3ObjectFileSize( size_t * pFileSize,
+                                         const TransportInterface_t * pTransportInterface,
+                                         const char * pHost,
+                                         const char * pMethod,
+                                         const char * pPath );
+
+/**
+ * @brief Send an HTTP request based on a specified method and path, then
+ * print the response received from the server.
+ *
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on success.
+ */
+static int uploadS3ObjectFile( const TransportInterface_t * pTransportInterface,
+                               const char * pHost,
+                               const char * pMethod,
+                               const char * pPath );
+
+/*-----------------------------------------------------------*/
+
+static int32_t connectToServer( NetworkContext_t * pNetworkContext )
```

Comment:
It's philosophical I was thinking more and it's weird to return EXIT_SUCCESS and EXIT_FAILURE from private functions. EXIT_SUCCESS and EXIT_FAILURE were meant for returning to the host OS that the program was successful. Perhaps you can follow the jobs demo and use bool or return the OpensslStatus_t.

Reply:
That makes sense, although in this case `connectToServer` is a paramater for the common util `connectToServerWithBackoffRetries`, so it matches the `TransportConnect_t` typedef signature (which returns EXIT_SUCCESS/FAILURE). So I suppose `http_demo_utils.h` along with all the files that already use it would have to be updated -- is that something that should be done here?

Reply:
Oh yes, I see. I don't want you to change code you didn't write. I will write a SIM thanks.